Sources/Auth.swift

```diff
@@ -0,0 +1,220 @@
+    let session = ASWebAuthenticationSession(
+      url: url,
+      callbackURLScheme: Constants.redirectUri.scheme!) { url, error in
+        if let error {
+          completion(.failure(error))
+          return
+        }
+
+        if let url {
+          let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+          let code = components!.queryItems!.first(where: { $0.name == "code"})
+          if let code {
+            self.exchangeAuthorizationCodeForTokens(authorizationCode: code.value!, verifier: verifier) { result in
+              completion(result)
+            }
+          } else {
+            completion(.failure(LoginError.noCode))
+          }
+        } else {
+          completion(.failure(LoginError.noUrl))
+        }
+      }
+    session.presentationContextProvider = PresentationContextProvider.shared
+    session.start()
+  }
+
+  private static func exchangeAuthorizationCodeForTokens(
+    authorizationCode: String,
+    verifier: String,
+    completion: @escaping (Result<AuthCodeResponse, Error>) -> Void)
+  {
+    let url = URL(string: "oauth/token", relativeTo: Constants.url)!
+
+    let parameters = [
+      "grant_type": "authorization_code",
+      "code_verifier": verifier,
+      "client_id": Constants.clientId,
+      "code": authorizationCode,
+      "redirect_uri": Constants.redirectUri.absoluteString,
+    ]
+
+    var request = URLRequest(url: url)
+    request.httpMethod = "POST"
+    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+    request.httpBody = try! JSONSerialization.data(withJSONObject: parameters, options: [])
+    
+    URLSession(configuration: URLSessionConfiguration.ephemeral).getAuthDataWith(request, completion: completion)
+  }
+
+  private static func getVerifier() -> String? {
+    let data = Data(count: 32)
+    var tempData = data
+    _ = tempData.withUnsafeMutableBytes {
+        SecRandomCopyBytes(kSecRandomDefault, data.count, $0.baseAddress!)
+    }
+    return tempData.a0_encodeBase64URLSafe()
+  }
+
+  private static func getChallenge(for verifier: String) -> String? {
+    guard let data = verifier.data(using: .utf8) else { return nil }
+
+    var buffer = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
+    _ = data.withUnsafeBytes {
+        CC_SHA256($0.baseAddress, CC_LONG(data.count), &buffer)
+    }
+    return Data(buffer).a0_encodeBase64URLSafe()
+  }
+
+  private static func generateDefaultState() -> String {
+    let data = Data(count: 32)
+    var tempData = data
+
+    let result = tempData.withUnsafeMutableBytes {
+        SecRandomCopyBytes(kSecRandomDefault, data.count, $0.baseAddress!)
+    }
+
+    guard result == 0, let state = tempData.a0_encodeBase64URLSafe()
+    else { return UUID().uuidString.replacingOccurrences(of: "-", with: "") }
+
+    return state
+  }
+}
+
+private class PresentationContextProvider: NSObject, ASWebAuthenticationPresentationContextProviding {
```

Comment:
We might want to make this configurable so users can pass in their own `ASWebAuthenticationPresentationContextProviding`